### PR TITLE
[Jormuangadr] Overriding fallback streetnetwork mode

### PR DESCRIPTION
**Pull request by xlqian**
https://jira.kisio.org/browse/NAVITIAII-3154

In this PR, we introduced a new concept: `overriding_mode`.

For example, if both `bss` and `walking` are given as `first_section_mode`, in order to remove duplicated journeys, we'll choose the 'best' fallback time for a given stop_point.

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py [hunk 1]
```diff
@@ -287,6 +288,25 @@ def __init__(
         self._request_id = request_id
         self._async_request()
 
+    @property
+    def _overriding_mode_map(self):
+        """
+        This map is used to storing the overriding modes of a give mode.
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py:294]:
```suggestion
        This map is used to store the overriding modes of a give mode.
```

File: source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py [hunk 2]
```diff
@@ -313,3 +333,22 @@ def wait_and_get(self, mode):
 
     def is_empty(self):
         return next((False for _, v in self._value.items() if v.wait_and_get()), True)
+
+    def get_best_fallback_durations(self, main_mode):
+        main_fb = self.wait_and_get(main_mode)
+        res = {}
+        overriding_modes = self._overriding_mode_map.get(main_mode, [])
+        overriding_fbs = [self.wait_and_get(mode) for mode in overriding_modes if self.wait_and_get(mode)]
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py:341]: I don't understand this line

File: source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py [hunk 3]
```diff
@@ -313,3 +333,22 @@ def wait_and_get(self, mode):
...
+
+        for uri, duration_item in six.iteritems(main_fb):
+            duration = duration_item.duration
+            # if the duration of the main_mode is smaller than all other mode, we keep the uri in the final res
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py:345]: :+1:

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py [hunk 4]
```diff
@@ -370,7 +375,7 @@ def _build_crowfly(pt_journey, entry_point, mode, places_free_access, fallback_d
 
     section_datetime = fallback_logic.get_pt_section_datetime(pt_journey)
     pt_datetime = fallback_logic.get_journey_bound_datetime(pt_journey)
-    fallback_duration = fallback_durations[pt_obj.uri].duration
+    fallback_duration = fallback_durations[pt_obj.uri]
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py:378]: why now this is not the duration field used

File: source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py (hunk 2 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/helper_classes/fallback_durations.py:341]: I am getting all fallback_durations of all overriding_modes

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py (hunk 4 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py:378]: fallback_durations is now a map of "uri" vs "duration"

**Review by pbench — APPROVED** [review]

**Review by benoit-bst — APPROVED** [review]

**Comment by xlqian** [comment]
Artemis OK!
https://ci.navitia.io/view/custom_artemis/job/custom_Artemis/41

**Pull request merged**